2024-09-26-typescript-effect/src/index.ts

```diff
@@ -0,0 +1,195 @@
+    apparent_temperature: Schema.String,
+    precipitation: Schema.String,
+  }),
+  current: Schema.Struct({
+    temperature_2m: Schema.Number,
+    relative_humidity_2m: Schema.Number,
+    apparent_temperature: Schema.Number,
+    precipitation: Schema.Number,
+  }),
+});
+
+type Weather = Schema.Schema.Type<typeof Weather>;
+
+// The field input
+const city = Option.fromNullable(
+  document.querySelector<HTMLInputElement>("#city"),
+);
+// The list of suggestions
+const cities = Option.fromNullable(
+  document.querySelector<HTMLUListElement>("#cities"),
+);
+// The weather information
+const weather = Option.fromNullable(
+  document.querySelector<HTMLDivElement>("#weather"),
+);
+
+const getRequest = (url: string) =>
+  // Effect.gen(function* () {
+  //   const client = yield* HttpClient.HttpClient;
+  //
+  //   return yield* pipe(
+  //     client.get(url),
+  //     HttpClient.withTracerPropagation(false),
+  //   );
+  // }).pipe(Effect.provide(FetchHttpClient.layer));
+  pipe(
+    HttpClient.HttpClient,
+    Effect.andThen((client) => client.get(url)),
+    HttpClient.withTracerPropagation(false),
+    Effect.provide(FetchHttpClient.layer),
+  );
+
+const getCities = function (search: string) {
+  Option.map(cities, (c) => (c.innerHTML = ""));
+
+  return pipe(
+    getCity(search),
+    Effect.map(renderCity),
+    // Check if the input is empty
+    Effect.when(() => Boolean(search)),
+  );
+};
+
+Option.map(city, (cityEl) => {
+  const stream = Stream.async(
+    (emit: StreamEmit.Emit<never, never, string, void>) =>
+      cityEl.addEventListener("input", function (_event) {
+        emit(Effect.succeed(Chunk.of(this.value)));
+      }),
+  );
+
+  pipe(
+    stream,
+    Stream.debounce(500),
+    Stream.runForEach(getCities),
+    Effect.runPromise,
+  );
+});
+
+const getCity = (city: string) =>
+  // Effect.gen(function* () {
+  //   const response = yield* getRequest(
+  //     `https://geocoding-api.open-meteo.com/v1/search?name=${city}&count=10&language=en&format=json`,
+  //   );
+  //
+  //   const geocoding = yield* pipe(
+  //     response,
+  //     HttpClientResponse.schemaBodyJson(GeocodingResponse), // Validate the response against the schema, it adds the `ParseResult.ParseError` to the `Error` type of the effect
+  //     Effect.orElseSucceed<GeocodingResponse>(() => ({ results: [] })), // If the effect is in a failure state, we can provide a default value to turn the effect into a success state
+  //   );
+  //
+  //   return geocoding.results;
+  // }).pipe(Effect.scoped);
+  pipe(
+    getRequest(
+      `https://geocoding-api.open-meteo.com/v1/search?name=${city}&count=10&language=en&format=json`,
+    ),
+    Effect.andThen(HttpClientResponse.schemaBodyJson(GeocodingResponse)),
+    Effect.orElseSucceed<GeocodingResponse>(() => ({ results: [] })),
+    Effect.map((geocoding) => geocoding.results),
+    Effect.scoped,
+  );
+
+const renderCity = (cities: readonly CityResult[]) => {
```

Comment:
Should be `renderCitySuggestions` to match the blogpost

Reply:
Done!